Autogenerate event pages

tovmasharrison opened the pull request:
# Description

The events can be autogenerated by adding WMT events to ```wmt_events.json``` and the rest to ```events.json```. I have tried to automate the most commonly added information about the events. However, specific information that is not automated can be added manually to the markdown file after generating the event. 

The ```_events.json``` and ```_wmt_events.json``` files include the empty structure of available fields that can be automated by choice. I have included them for reference. 

Also, I have created a list to show all the MT Summit, EAMT, and AMTA events separately at the top of the page. This closes that part in #584.

How it works:
- **Validation**: After the events are added to the corresponding ```.json``` file, ```generate.py``` validates the entries and creates a file inside the ```events/``` directory. For external events or for events where there's no need to create a page, the ```"id"``` key-value pair should not be added. The examples are **AMTA 2024**, **AmericasNLP** and **WMT24**.

- **```events.md```**: An appropriate header is created based on the event's ```startDate```, and each event is listed under the corresponding header appropriately. For example, if the event starts in 2024 or 2025, it will be listed under **2024 Events** or **2025 Events**. 

- **```calls-for-papers.md```**: If the event has a call for paper, then ```"callsForPapersDeadline": "some_date"```  key value pair should be given inside the ```.json``` file for it to be added to ```calls-for-papers.md``` or else it won't. 

- **Bold/unbold dates and names based on the deadline**: The names and dates are in bold if the deadline has not been passed, and as soon as it passes they will automatically unbold. The check to bold/unbold will happen each time the website is generated, as it will compare the ```endDate``` with the current date. 

- **Past/future tense**: When adding the event to the ```.json``` file, the future and past tenses can be added simultaneously. For the future tense, it can be added as ```"openingParagraph": "The event will take place"```, for the past tense, it can be added as ```"pastOpeningParagraph": "The event took place"```, and the appropriate tense will be automatically shown with the same bold/unbold logic mentioned above. 

- **Syntax**: The markdown syntax should be replaced with HTML tags inside the ```.json``` file. 
                    \<strong>Some word\</strong> should be used to bold the name.
                    \<a href='/link'>some_word\</a> should be used to add a link.


Please let me know if you have any questions or suggestions. 

Fixes #563 

### Checklist:

- [x] I have read the [contributing guidelines](/CONTRIBUTING).
- [x] I have followed the [style guide](http://machinetranslate.org/style).

bittlingmayer commented on generate.py:
```diff
@@ -724,3 +733,185 @@ def normalize(code):
+    **event
+  }
+
+  # Doesn't create a file if it's an external event, as external events should not have an id
```
What's an example of an "external" event?

tovmasharrison replied:
For example, ```AmericasNLP```. We don't have a page for it but we link the event externally.

bittlingmayer commented on generate.py:
```diff
@@ -724,3 +733,185 @@ def normalize(code):
+  name = event['name']
+
+  # Start and end dates are necessary for listing the events on events.md
+  if not isinstance(event['startDate'], str):
```
Really we should be checking that it's parsed as a date.

bittlingmayer commented on generate.py:
```diff
@@ -724,3 +733,185 @@ def normalize(code):
+  if not isinstance(event['startDate'], str):
+    raise Exception(event['startDate'])
+
+  if not isinstance(event['endDate'], str):
```
Maybe if endDate is missing, we assume it's just a single day event?

tovmasharrison replied:
Yes we can

bittlingmayer commented on _data/_events.json:
```diff
@@ -0,0 +1,239 @@
+      "location":"",
+      "startDate":"",
+      "endDate":"",
+      "date":"",
```
Do we need this in addition to startDate?

tovmasharrison replied:
I added a separate ```date``` for cases where we specify the date in words. 
![Screenshot from 2024-01-05 11-16-51](https://github.com/machinetranslate/machinetranslate.org/assets/94985882/5e93a986-8d83-47fc-a624-0f9cfc73f039)

tovmasharrison replied:
Hi!

I'll take a look.

tovmasharrison replied:
Hi @cefoo, 

Since we are auto-generating the events, there wasn't a straightforward way to place the WMT24 event manually at the top. However, I provided an approximate date for it based on the rest of the WMT events so it can be at the top until the actual date becomes known. 

Please let me know if that solution is ok.

tovmasharrison replied:
Edit: I actually added it at the very top.

cefoo replied:
@liashahnazaryan has just created the WMT24 page, with date on December.

However, it would be good to enforce WMT to be on top in future years, so this could be a good practice moving forward. WMT is usually on December, I guess we could just put December 31st until we know the real date. However, that date should be hidden until we know the real one.

bittlingmayer commented on _data/_events.json:
```diff
@@ -0,0 +1,239 @@
+      "linkUrl":[
+         ""
+      ],
+      "impDatesHeader":"",
```
What is this?

If `imp` stands for "important", let's just write `importantDatesHeader`.

bittlingmayer replied:
I think we just want:
```
"importantDates": [ {
        "name": "...",
        "date": "..."
    },
    ...
]
```

bittlingmayer commented on _data/_events.json:
```diff
@@ -0,0 +1,239 @@
+         ""
+      ],
+      "blockquoteSpeakersHeader":"",
+      "speakHeadcontent":[
```
`Content` with uppercase

bittlingmayer commented on _data/events.json:
```diff
@@ -0,0 +1,234 @@
+       "location":"Chicago, Illinois"
+    },
+    {
+       "name":"<a href='https://turing.iimas.unam.mx/americasnlp/2024_workshop.html'>AmericasNLP</a>",
```
I don't think there should be a URL in the name.

bittlingmayer commented on _data/_events.json:
```diff
@@ -0,0 +1,239 @@
+   {
+      "name":"",
+      "id":"",
+      "title":"",
```
Do we need both name and title?

bittlingmayer commented on _data/_events.json:
```diff
@@ -0,0 +1,239 @@
+            "date":""
+         }
+      ],
+      "bulletKeyNoteSpeakOrTopsHeader":"",
```
Again, we should have this.

Instead of ""bulletKeyNoteSpeakOrTopsHeader" and "names", we need just

```
"organisers": [
   {
       "name": "...",
       "institution": "..."
   },
   ...
]
```

bittlingmayer commented on _data/_events.json:
```diff
@@ -0,0 +1,239 @@
+            "url":""
+         }
+      ],
+      "sharedTasksHeader":"",
```
Isn't this only for WMT?

bittlingmayer commented on _data/_events.json:
```diff
@@ -0,0 +1,239 @@
+            "optionalSentence":""
+         }
+      ],
+      "seo":{
```
This should not be in JSON.

It should get generated in markdown.

bittlingmayer commented:
The JSON schema needs to be way more minimal.

We want to reduce the burden on people adding events.

It's fine if we harmonise a bit how events are displayed.

bittlingmayer commented:
Having a JSON schema is a good idea.

We should just use something like https://json-schema.org/, and then we can quickly run a validator.

Instead of using ad-hoc schema format, and then still needing dozens of if-statements.

bittlingmayer commented on _data/_wmt_events_schema.json:
```diff
@@ -0,0 +1,253 @@
+            "items":{
+              "type":"object",
+              "properties":{
+                "language":{
```
do we need `language` if we have `language_pair`?

tovmasharrison replied:
I think so because the table is for one language but with multiple language pairs. 
![Screenshot from 2024-02-08 15-26-37](https://github.com/machinetranslate/machinetranslate.org/assets/94985882/f0e80267-a049-4c83-b9d9-d2f19e678be4)

In the above screenshot, the language is English, and the language pairs are inside the Language Pairs columns. 

What do you think?

bittlingmayer commented on _data/_wmt_events_schema.json:
```diff
@@ -0,0 +1,253 @@
+                "language_pair":{
+                  "type":"string"
+                },
+                "System":{
```
should be all lowercase

bittlingmayer commented on _data/events.json:
```diff
@@ -0,0 +1,321 @@
+         "type":"Person",
+         "url":"http://turing.iimas.unam.mx/americasnlp/st.html"
+      },
+      "futureTenseOpeningParagraph": "The <strong>Fourth AmericasNLP Competition</strong> will take place online in June, at NAACL 2024 in Mexico City. The competition focused on creating machine translation systems for indigenous languages from the Americas.",
```
Ideally we'd avoid this.

Also should we use Markdown instead of HTML?

tovmasharrison replied:
Should we just remove future/past tenses and only keep ```opening_paragraph```? But this way, it should be modified manually when the event happens.

Well, each event page is rendered through HTML. I don't think there is a straightforward way to use Markdown syntax inside HTML?

bittlingmayer commented on _data/events.json:
```diff
@@ -0,0 +1,321 @@
+      "speakers": [
+         {
+            "type": "Keynote speakers",
+            "about": [
```
feels wrong

i'd expect something more like this:

```
"speakers": [
       { name: "Graham Neubig", type: "keynote" },
       { name: "Jaime Pérez González", type: "keynote" },
       "Manuel Mager",
       "Abteen Ebrahimi",
       "Shruti Rijhwani",
       "Arturo Oncevay",
       "Luis Chiruzzo",
       "Robert Pugh",
       "Katharina Kann"
]
```

though not sure if the schema will allow either string or object

tovmasharrison replied:
The schema won't allow this since an array should contain elements of the same type.

Probably something like this can do:

```"speakers": [
    { "name": "Graham Neubig", "type": "keynote" },
    { "name": "Jaime Pérez González", "type": "keynote" },
    { "name": "Manuel Mager" },
    { "name": "Abteen Ebrahimi" },
    { "name": "Shruti Rijhwani" },
    { "name": "Arturo Oncevay" },
    { "name": "Luis Chiruzzo" },
    { "name": "Robert Pugh" },
    { "name": "Katharina Kann" }
]```

@bittlingmayer, What do you think?

bittlingmayer commented on _data/events.json:
```diff
@@ -0,0 +1,321 @@
+      "importantDates": [
+         {
+            "name": "Start of the anonymity period",
+            "date": "10 February"
```
dates should always be machine readable and include the year, so "2024-02-10"

i think we can also make this simpler:

"importantDates": {
   "Start of the anonymity period": "2024-02-10",
   ...
}

bittlingmayer commented on _data/events.json:
```diff
@@ -0,0 +1,321 @@
+         "url":"https://iwslt.org/2023/#organizers"
+      },
+      "futureTenseOpeningParagraph":"The International Conference on Spoken Language Translation (<strong>IWSLT</strong>) is an annual scientific conference on spoken language translation.<br> <br> IWSLT 2024 will be co-located at ACL 2024 online and in Bangkok, Thailand from 15 to 16 August, 2024. <br><br> IWSLT 2024 will be a hybrid event.",
+      "pastTenseOpeningParagraph":"The nineteenth Machine Translation Summit 2023 (<strong>MTS 2023</strong>) took place from 4 September to 8 September in Macau Special Administrative Region, China. MTS 2023 was organised by [AAMT](/aamt).",
```
these 2 paragraphs are not the same, the past one has a different year and mixes in Markdown.

bittlingmayer commented on _data/wmt_events.json:
```diff
@@ -0,0 +1,212 @@
+     "multidaySchedule":[
+       [
+         {
+           "time":"10:30 - 11:00",
```
I think we should keep it machine readable

So `startTime`, and then an optional `endTime`.

Or even just
```
[
    "10:30": "Some activity"
]
```
No end times.

bittlingmayer commented on languages/emj.md:
```diff
@@ -0,0 +1,22 @@
+  name: Yandex Translate
+supported_qe_apis: []
+seo:
+  name: Machine translation for None
```
Seems like a bug

tovmasharrison replied:
Yes, I'll fix that with a separate PR.

tovmasharrison replied:
Fixed in #598.

bittlingmayer commented on index/index.md:
```diff
@@ -57,8 +57,8 @@ The **Machine Translate Foundation** is building **open information** and **comm
 > #### Featured events
-> - [AMTA 2024](/amta2024) - 30 September-02 October, Chicago 🇺🇸
-> - [EAMT 2024](/eamt2024) - 24-27 June, Sheffield 🇬🇧
+> - [AMTA 2024](/amta-2024) - 30 September-02 October, Chicago 🇺🇸
```
But amta2023 we still have without dash.  Probably should all be with dash, to represent the space.

bittlingmayer commented:
Yes, with schema is better!

Made a few comments, but today is busy with the community meetup, will look more on the weekend.

bittlingmayer commented on generate.py:
```diff
@@ -44,6 +49,30 @@
 with open ('_data/quality_estimation.json', 'r', encoding='utf-8') as stream:
   QUALITY_ESTIMATION = json.load(stream)
 
+### Read events
```
I'd probably do something more like this

```
### Read events and events schema
with open ('_data/events.json', 'r', encoding='utf-8') as f:
  EVENTS = json.load(f)

with open('_data/_events_schema.json', 'r', encoding='utf-8') as f:
  EVENTS_SCHEMA = json.load(f)

### Read WMT events and WMT events schema
...
```

bittlingmayer commented on events/wmt-test.md:
```diff
@@ -0,0 +1,162 @@
+  scoring_table:
+  - language: ->English
+    sentence: Lorem Ipsum is simply dummy text of the printing and typesetting industry.
+    language_pair: English <-> Czech
```
let's use Unicode, e.g. `↔`

bittlingmayer commented on _layouts/wmt_event.html:
```diff
@@ -0,0 +1,223 @@
+
+{% assign current_date = site.time | date: "%Y-%m-%d" %}
+
+<p>{% if current_date < page.end_date %}{{ page.future_tense_opening_paragraph }}{% else %}{{ page.past_tense_opening_paragraph }}{% endif %}</p>
```
Really we should do this in JavaScript so that we don't need to rebuild the site for it to be correct.

bittlingmayer commented on _data/_wmt_events.json:
```diff
@@ -0,0 +1,161 @@
+      "name": "Name of the organizer",
+      "url": "URL about the organizer"
+    },
+    "future_tense_opening_paragraph": "An opening paragraph in future tense about the event, when/where it's happening.",
```
Can we just autogenerate these in the .md?

For example:

```

dates = f'on { start_date }' if end_date == start_date else f'from { start_date } to { end_date }'

location = 'online' if ... else f'in { location_name }'

future_tense_opening_paragraph = f'{ name } will take place { dates} { location }.'

future_tense_opening_paragraph = f'{ name } took place { dates} { location }.'

```

And then `summaries` can be renamed `more_paragraphs`.

bittlingmayer commented on events/iwslt-2024.md:
```diff
@@ -0,0 +1,77 @@
+  - Marcello Federico, AWS AI Labs, United States
+  - Alex Waibel, CMU, United States
+  - Jan Niehues, KIT, Germany
+  - "Sebastian St\xFCker, KIT, Germany"
```
Can't we use `ü`?

bittlingmayer commented on events/iwslt-2024.md:
```diff
@@ -0,0 +1,77 @@
+  - Elizabeth Salesky, JHU, United States
+  - Atul Kr. Ojha, University of Galway, Ireland
+shared_tasks:
+- <a href='https://iwslt.org/2024/s2s'>Speech-to-speech track</a>
```
This is hacky.

We should do either like this:

```
shared_takes:
- name: Speech-to-speech track
   url: https://iwslt.org/2024/s2s
...
```

Or like this:
```
shared_takes:
- Speech-to-speech track: https://iwslt.org/2024/s2s
...
```

In JSON the equivalent would be:
```
"shared_tasks": [
      {
        "name": "Speech-to-speech track",
        "url": "https://..."
      }
      ...
  ]
```
or
```
"shared_tasks": {
      "Speech-to-speech track": "https://..."
     ...
}
```

bittlingmayer commented on _data/_events.json:
```diff
@@ -0,0 +1,117 @@
+     "speakers": [
+       {
+         "type": "Type of the speaker: Panellist/Invited/Organizer",
+         "about": [
```
This is clunky.

bittlingmayer replied:
```
"speakers": {
  "Organizers": [ ... ]
  ...
}
```
(and let them use whatever keys)

tovmasharrison commented:
@bittlingmayer All done.

bittlingmayer commented on _data/events.json:
```diff
@@ -0,0 +1,235 @@
+      "location":{
+         "type":"VirtualLocation",
+         "name":"Online",
+         "event_attendance_mode": "OnlineEventAttendanceMode"
```
this is way too verbose, 3 lines to basically just say that it's online.

bittlingmayer commented on _data/events.json:
```diff
@@ -0,0 +1,235 @@
+         "event_attendance_mode": "OfflineEventAttendanceMode"
+      },
+      "organizer":{
+         "type":"Person",
```
not a person

tovmasharrison replied:
https://iwslt.org/2023/#organizers

It looks like the organisers of this event are individual people and not an organisation/association?

bittlingmayer commented on _data/events.json:
```diff
@@ -0,0 +1,235 @@
+      "important_dates":
+         {
+            "Release of shared task training and developer data":"2024-01-15",
+            "Evaluation":"1 - 15 April",
```
just put start date

bittlingmayer commented on _data/events.json:
```diff
@@ -0,0 +1,235 @@
+            "Notification of acceptance":"2024-06-04",
+            "Camera ready deadline":"2024-06-24",
+            "Pre-recorded video due":"2024-07-22",
+            "Conference":"15 - 16 August"
```
this is at the top, so not needed

bittlingmayer commented on _data/events.json:
```diff
@@ -0,0 +1,235 @@
+      "start_date":"2024-07-03",
+      "end_date":"2024-07-06",
+      "calls_for_papers_deadline":"2024-03-31",
+      "location":{
```
I think for location let's just do

```
"location": "Varna, Bulgaria"
```
and 
```
"online": true
```

A mixed event is just an event where both location and online are set to something.

or we can even do

```
"location": "Varna, Bulgaria and online"
```

bittlingmayer commented on _data/events.json:
```diff
@@ -0,0 +1,235 @@
+      },
+      "start_date":"2024-06-24",
+      "end_date":"2024-06-27",
+      "links":{
```
just make it a list
we can remove the https:// programmatically for display

bittlingmayer commented on _data/events.json:
```diff
@@ -0,0 +1,235 @@
+         },
+      "shared_tasks":[
+         {
+         "name": "Speech-to-speech track",
```
fix indenttion

bittlingmayer commented on _data/events.json:
```diff
@@ -0,0 +1,235 @@
+      "organizer":{
+         "type":"Organization",
+         "name":"NeTTT",
+         "url":"http://machinetranslate.org/nettt2022.md#organisers"
```
this can't be right

bittlingmayer commented on _data/events.json:
```diff
@@ -0,0 +1,235 @@
+      },
+      "important_dates":
+         {
+            "Deadline for paper submission":"2024-03-08",
```
space after `:`

bittlingmayer commented on _data/events.json:
```diff
@@ -0,0 +1,235 @@
+            "Notification to paper authors":"2024-04-08",
+            "Camera ready deadline":"2024-04-22",
+            "Author Registration":"2024-05-08",
+            "Paper Conference":"24 - 27 June",
```
start date only, in machine readable format

bittlingmayer commented on _data/_events_schema.json:
```diff
@@ -0,0 +1,166 @@
+{
+  "$schema":"http://json-schema.org/draft-07/schema#",
+  "$id":"https://machinetranslate.org/events-schema",
+  "title":"Events",
```
somehow we're missing URL of the event

bittlingmayer commented on _data/events.json:
```diff
@@ -0,0 +1,235 @@
+      }
+   },
+   {
+      "name": "<a href='/meetup'>Machine translation meetup</a>",
```
every event should have a url

in this case:
```
"url": "https://machinetranslate.org/meetup",
```

tovmasharrison commented:
@bittlingmayer Done

bittlingmayer commented on _data/_events.json:
```diff
@@ -0,0 +1,118 @@
+       "url": "URL for the location where the event will happen"
+     },
+     "organizer": {
+       "type": "Association/Organization/etc.",
```
I think we can remove this.

tovmasharrison commented:
@bittlingmayer 

I have added a form submission at ```http://127.0.0.1:4000/add_event``` that returns a JSON version of the inputs, which can be copied and pasted inside the ```events.json``` / ```wmt_events.json``` / ```calls_for_papers.json``` files. 

I have reverted the structure for ```speakers```, ```important_dates```, ```multiday_schedule```, and ```one_day_schedule``` to their previous states since their input boxes were being represented as textarea. See below.
<img width="889" alt="Screen Shot 2024-02-22 at 13 53 20" src="https://github.com/machinetranslate/machinetranslate.org/assets/94985882/c6183674-d68f-4461-916f-2ae83d2c1efb">


Also, the ```WMT Test`` event is still included for reference. I'll remove it once everything looks good. 

Let me know what you think.

bittlingmayer commented on events/eamt-2024.md:
```diff
@@ -1,75 +1,76 @@
     type: Place
     name: Sheffield, England
-
+    url: null
```
It's up now: https://eamt2024.sheffield.ac.uk/

bittlingmayer commented on events/americasnlp-2024.md:
```diff
@@ -0,0 +1,82 @@
+  location:
+    type: Place
+    name: Mexico City, Mexico
+    url: null
```
Can we get rid of these?  We already have links at the event level.

bittlingmayer commented on events/americasnlp-2024.md:
```diff
@@ -0,0 +1,82 @@
+  organizer:
+    type: Person
+    name: null
+    url: http://turing.iimas.unam.mx/americasnlp/st.html
```
This is more like the Shared Task URL.  Not the organizer.

bittlingmayer replied:
It's say the organizer is http://turing.iimas.unam.mx.

So Organization not Person.

bittlingmayer commented on events/eamt-2024.md:
```diff
@@ -1,75 +1,76 @@
+  type: Organization
+  name: European Association of Machine Translation
+  url: https://eamt.org
+start_date: '2024-06-24'
```
I think we can print these without quotes.

bittlingmayer replied:
Only if it is easy - I know it is probably just the default in the conversion to YAML.

bittlingmayer commented on events/add_event.md:
```diff
@@ -0,0 +1,5 @@
+---
+layout: "default"
```
Should be:

---
layout: "add_event_form"
---

and no `include`

bittlingmayer commented on _includes/form.html:
```diff
@@ -0,0 +1,656 @@
+<head>
+  <meta charset="utf-8" />
+  <title>Enter Event Details</title>
+  <link rel="stylesheet" style="text/css" href="/json_form/deps/opt/bootstrap.css" data-proofer-ignore />
```
Let's just do like this so we don't have to host these files ourselves:

https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/css/bootstrap.min.css

https://www.w3schools.com/bootstrap/tryit.asp?filename=trybs_default&stacked=h
